Enhancements for shell plugins

bfloch opened the pull request:
This fixes and improves the shell plugins, especially on Windows for cmd and PowerShell-like.
Formally excluded shell dependent tests are now passing.

I would love to have someone run the test-suite on MacOS.

### Future work
- Better Visual Studio handling in CMake for Windows
- Remove dependency of pwsh on SH-class for `get_syspaths` under Linux

### Fixes
- #694 Expand both generic environment variables `$ENVVAR / ${ENVVAR}` as well as shell specific ones correctly, like `%ENVVAR% ${Env:ENVVAR}` for cmd and PowerShell-like shells
- #691 Undoes regression in quotes handling of cmd

### Features / Enhancements
- Added `get_all_key_tokens` to allow for multiple key-forms
- Added `convert_tokens` to normalize token forms to shell specific form
- Allow pwsh shell on Linux (and pass tests)
- Implements optional script execution on windows via PATHEXT.
  - Windows shells may add custom PATHEXT values in plugin config file
  - `create_executable_script` now returns list of created scripts based on `config.create_executable_script_mode`:
     - `single` (default) Just create the requested script. Usually extension-less.
     - `py` create .py script that will allow launching scripts on windows,
     - `platform_specific` will create .py script on windows and requested on other platforms
    - `both` creates the requested file and a .py script for windows

- Introduced Shell specific regex parsing to handle ambiguous case like `set PATH=$ENV:VAR` in PowerShell vs. Bash within the NamespaceFormatter
- Introduced printout of shell name for `shell_dependent` test failure

### Tests passing
- [x] Windows 10.0.17134
  - [x] cmd*
  - [x] PowerShell 5.1.17134.165*
  - [x] pwsh 6.2.2*
  - *fails in `test_build_cmake` like in master. Will be addressed next.
- [x] Linux Ubuntu 18.04.2 LTS (WSL)
  - [x] tcsh
  - [x] sh
  - [x] bash
  - [x] csh
  - [x] pwsh 6.2.1
  - [x] zsh ~~fails like on master~~ Passes based on #711

### Backwards compatibility
- Fully backwards compatible, new features use config options default to current behaviour
- Shell Plugin API has been extended slightly.
  - Implement `get_all_key_tokens(key)` (with safest/common form at index 0)
  - Implement `line_terminator()` for shell/platform specific line_terminator as expected by the shell
  - `join` now a class method
  - Extend `ENV_VAR_REGEX` if needed with shell specific token forms
  - You can make use of new `convert_tokens` in you implementation of `info` and `error`

nerdvegas commented on src/rez/rex.py:
```diff
@@ -891,7 +896,11 @@ def escape_envvar(matchobj):
-        format_string_ = re.sub(self.ENV_VAR_REGEX, escape_envvar, format_string)
+        regex = ActionInterpreter.ENV_VAR_REGEX
+        if "regex" in kwargs:
+            regex = kwargs["regex"]
```
`regex = kwargs.get("regex") or ActionInterpreter.ENV_VAR_REGEX`

nerdvegas commented on src/rez/rezconfig.py:
```diff
@@ -619,6 +619,29 @@
 
 
+###############################################################################
+# Utils
```
I would just add this to the `Misc` section, it's not that clear what "utils" really means in the context of rez settings.

nerdvegas commented on src/rez/shells.py:
```diff
@@ -160,6 +160,29 @@ def spawn_shell(self, context_file, tmpdir, rcfile=None, norc=False,
         raise NotImplementedError
 
+    @classmethod
+    def get_key_token(cls, key, form=0):
```
I find the use of `get_key_token`/`key_form_count` a bit confusing. How about we just update `get_key_token` to return a list, where the first item is the preferred form? It's not used in many places so not much code would need updating, and I'd be pretty surprised if anyone considered this part of the public API (so I don't expect backwards compatibility to be an issue here).

nerdvegas replied:
To add to this - the func should be allowed to return str or list of str, so that any existing external shell plugins won't break. So in that respect we need backwards compatibility, but I think that will be enough.

bfloch replied:
Did it as I proposed. This is way cleaner, the API stays compatible and the naming is way clearer.
Hope you agree.

nerdvegas commented on src/rez/tests/test_shells.py:
```diff
@@ -69,27 +63,69 @@ def test_no_output(self):
         if command:
             r = self._create_context(["hello_world"])
-            p = r.execute_shell(command="hello_world",
+            script = "hello_world"
```
`script` as a var here is a bit odd?

nerdvegas commented on src/rez/tests/test_shells.py:
```diff
@@ -145,13 +181,19 @@ def test_rcfile(self):
+        # TODO: If this exception was created for the following shells then
+        #  it is redundant. Can we clarify which platforms this was meant for?
+        has_buildin_echo = sh.name() in ['cmd', 'powershell', 'pwsh']
+        if not echo_cmd and not has_buildin_echo:
```
Could we keep this simple and just use whatever the `echo` equivalent is when on Windows? Are there any non-windows shells that wouldn't have `echo`?

nerdvegas commented on src/rez/tests/test_shells.py:
```diff
@@ -189,21 +231,29 @@ def _execute_code(func, expected_output):
             self.assertEqual(p.returncode, 0)
-            token = '\r\n' if platform_.name == 'windows' else '\n'
+
+            # Powershell and Unix uses \n
```
Perhaps it would be better to add a `line_terminator` property to Shell class, rather than hardcoding this behaviour here. I had probably assumed at the time that terminator was platform-dependent and not shell-dependent.

bfloch replied:
I've added a `line_terminator` to the shell API.

nerdvegas commented on src/rez/tests/test_shells.py:
```diff
@@ -189,21 +231,29 @@ def _execute_code(func, expected_output):
+                if is_powershell:
+                    info('`"{}`"'.format(sh.get_key_token("FOO")))
+                else:
+                    info('"{}"'.format(sh.get_key_token("FOO")))
```
So the original code here should really just have been:

    info('"${FOO}"')

The test is showing two things - that ${X}/$X variable references are getting correctly translated to their shell equivalent; and that use of expandable/literal is working as expected. Rather than directly use `get_key_token` anywhere here, the idea was just to test the actual output from the shell.

Wrt the whole adding of double quotes here to avoid problems with '>' etc, it does seem like that should be dealt with by `info` itself (and by shell-specific string expansion in general). But that stuff gets tricky and a change related to that should be in a separate PR. Is Powershell a special case where this has to be done differently?

bfloch replied:
This now passes and I was able to do the right thing by introducing `convert_tokens` to Shell.
This was actually simpler that both of us thought 👍

nerdvegas commented on src/rez/tests/test_shells.py:
```diff
@@ -225,19 +275,32 @@ def _print(value):
             _print('hey < world')
             _print('hey ^ world')
 
+            # Platform dependent form of variables.
```
I would just include unix shells here also, rather than have an exception at all. It should work also and it doesn't really matter if it re-tests some of the same things.

bfloch replied:
Removed the platform checks. Way cleaner, thanks.

nerdvegas commented on src/rez/tests/test_shells.py:
```diff
@@ -294,6 +372,25 @@ def _rex_appending():
         _execute_code(_rex_appending, expected_output)
 
+    @shell_dependent()
+    def test_variable_encoding(self):
```
I think we could do without this test actually - these tests are really only asserting internal implementation details of the shells. If you go down this road (eg, testing that powershell class acts like powershell) you'd have to write loads more tests, and I don't really see the value in that. I would argue that that isn't what this test suite is for - it's for checking that rex code (which is shell agnostic by design) gives the expected result in all shells - no more.

bfloch replied:
You are right.

nerdvegas commented on src/rez/util.py:
```diff
@@ -10,11 +10,35 @@
+    # Start with 1 to not collide with None checks
+
+    # Requested shebang script only. Usually extension-less.
+    requested = 1
```
I don't really understand what is meant by "requested" here? And it isn't clear what "requested" would mean when on Windows... isn't it the same as "platform_specific"?

Would "native" be a better term? I'm trying to understand what "requested" means as in the term by the way, not the implementation!

bfloch replied:
I renamed it to "single". Hope that makes more sense.
I left `platform_specific`, but it you prefer `native` let me know.

nerdvegas commented on src/rez/util.py:
```diff
@@ -46,18 +82,72 @@ def create_executable_script(filepath, body, program=None):
+        platform (str): Platform to evaluate the script files for
+
+    Returns:
+
```
Missing docstring

nerdvegas commented on src/rezplugins/shell/cmd.py:
```diff
@@ -7,7 +7,7 @@
 from rez.utils.system import popen
 from rez.utils.platform_ import platform_
-from rez.util import shlex_join
+from rezplugins.shell.powershell import list2cmdline
```
This is nasty! If there's common functionality here then it should be moved into util code somewhere. There certainly should not be cross referencing of shell implementations like this

nerdvegas commented on src/rezplugins/shell/powershell.py:
```diff
@@ -1,39 +1,54 @@
     @classmethod
     def name(cls):
-        return 'powershell'
+        raise NotImplementedError
```
just drop this method, no need to implement here

nerdvegas commented on src/rezplugins/shell/powershell.py:
```diff
@@ -1,39 +1,54 @@
     @classmethod
     def file_extension(cls):
-        return 'ps1'
+        raise NotImplementedError
```
just drop this method, no need to implement here

nerdvegas commented on src/rezplugins/shell/pwsh.py:
```diff
@@ -1,25 +1,12 @@
-from subprocess import PIPE
-import os
-import re
+from rezplugins.shell.powershell import PowerShellBase
```
Yeah I see that we need to do this here

bfloch replied:
I've added a powershell_common module.

nerdvegas commented on src/rezplugins/shell/pwsh.py:
```diff
@@ -1,25 +1,12 @@
-import os
-import re
+from rezplugins.shell.powershell import PowerShellBase
+from rezplugins.shell.sh import SH
```
..but this should be changed

bfloch replied:
Hope that comment makes it clearer why we use SH here.

nerdvegas commented on src/rez/bind/hello_world.py:
```diff
@@ -43,7 +43,15 @@ def bind(path, version_range=None, opts=None, parser=None):
         filepath = os.path.join(binpath, "hello_world")
-        create_executable_script(filepath, hello_world_source)
+
+        # In order for this script to run on each platform we create the
```
While this comment is correct, I don't think it belongs here. This caveat is already pointed out in a very similar comment with the function definition.

bfloch commented:
Thanks for the review. I'll dropped notes/questions and will look into the fixes.

bfloch commented:
I've updated the first post and addressed everything in your review.
One thing I could not solve is the SH reference in pwsh, but I limited the scope and added a proper comment with a reference how this might be solved upstream of pwsh.

Ran the tests again, pass as before.

maxnbk commented:
FYI, zsh tests on master should be fixed when #711 is merged in.
Notifying in case it matters for this PR.

nerdvegas commented on src/rez/shells.py:
```diff
@@ -160,7 +176,46 @@ def spawn_shell(self, context_file, tmpdir, rcfile=None, norc=False,
+        return cls.get_all_key_tokens(key)[0]
+
+    @classmethod
+    def get_all_key_tokens(cls, key):
```
:+1:

nerdvegas commented:
I'm ready to merge this, but would like to hear RE OSX testing also. In the meantime I'm going to start looking at GitHub Actions, so we'll have this testing covered anyway (and across py2/3 also, which is going to become important pretty soon).

nerdvegas commented on src/rezplugins/shell/cmd.py:
```diff
@@ -182,12 +184,34 @@ def _create_ex():
 
+        # Make .py launch within cmd without extension.
+        if self.settings.additional_pathext:
+            executor.command('set PATHEXT=%PATHEXT%;{}'.format(
```
It's a fairly minor point, but in nested rez-envs this will cause PATHEXT to become `>PY;.PY;.PY...`. If it's easy, it may be worth updating this command to only add the additional pathext if it isn't already in PATHEXT. If it's a pain I wouldn't worry too much.

nerdvegas replied:
Actually, note also that wrt #707 (see my comments there), that with sufficient env-var control added to the project, this special case may not be necessary at all. I'm imagining something like:

```
env_var_settings = {
  "PATHEXT": {
    "set_always": ".PY",  # always set PATHEXT=.PY in a new shell...
    "path_modify_mode": "retain_after"  # ..but also keep pre-rez value of PATHEXT appended as well
  }
}
```

bfloch replied:
I personally can wait to have #707 done first. Everyone who has a need for better shell support may feel free to merge this branch for testing.
Up to you - I will fix the issue above as suggested, just in case.

bfloch replied:
For now I've implemented de-duplication as is for cmd and PowerShell.
I will move the discussion about the generic variable handling to the #707

bfloch commented:
> FYI, zsh tests on master should be fixed when #711 is merged in.
> Notifying in case it matters for this PR.

Cool will try merging this in. Thanks!

bfloch commented:
Will fix the conflict and I'll give the de-duplication on cmd another shot today.
Might have to do with shell=False again, not sure. The syntax certainly works within cmd sessions ...

bfloch commented:
> FYI, zsh tests on master should be fixed when #711 is merged in.
> Notifying in case it matters for this PR.

Notable changes since your last review:
- Fixed promp not expanded with cmd (probably because variable expansion now works properly)
- Supressing PowerShell header
- Added PATHEXT deduplication for cmd and PowerShell as suggested
- Rebased on current master. Fixed the zsh failure as advertised. Thanks.

So now it is just a question if we want #707 first.
I guess since it technically does not rely on it and we do have quite a lot of changes here we should probably go ahead and merge it. At least we get some real feedback early on, while there still might be some design discussions happening in #707.

All tests (except cmake related) pass on Win & Linux.
MacOS testing is dependent on the github actions setup, I presume. I'll leave that part to you guys since I don't have access to a Mac.

nerdvegas commented:
Thanks Blazej. I'm happy to merge this and will do soon - it stands on its own from #707 and fixes an outstanding bug on Windows.

nerdvegas commented:
Just an update - getting to this soon. We're near the end of the py3 work, I'm hoping to get that done first then merge this right after.

bfloch commented:
No worries, there is no rush with this merge. I would love to hear your feedback on #737 though since this is the last missing piece for our Rez deployment.

Btw. I haven't tested this PR with python 3, so let me know if there is any work to be done.
In fact, we as a Blender-based animation studio would love to run rez with python 3 so I might be running tests based on your effort soon.

nerdvegas commented:
Yeah I'm pretty slammed atm but I'm keen to get onto both #737 and #703 so
I'll do that soon.

We're maybe a few days/a week away from wrapping up on py3 port, so I
reckon if you could do some testing on it once that's done, that would be
very helpful.

Cheers
A


On Wed, Sep 11, 2019 at 11:03 AM Blazej Floch <notifications@github.com>
wrote:

> No worries, there is no rush with this merge. I would love to hear your
> feedback on #737 <https://github.com/nerdvegas/rez/issues/737> though
> since this is the last missing piece for our Rez deployment.
>
> Btw. I haven't tested this PR with python 3, so let me know if there is
> any work to be done.
> In fact, we as a Blender-based animation studio would love to run rez with
> python 3 so I might be running tests based on your effort soon.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nerdvegas/rez/pull/698?email_source=notifications&email_token=AAMOUSRPVVQH3FXG5UHV6R3QJA7VBA5CNFSM4IPRPML2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6M5SBY#issuecomment-530176263>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMOUSQHJLGVHIA4USJAXMTQJA7VBANCNFSM4IPRPMLQ>
> .
>
